bevy_picking release notes (#1806)
bevyengine/bevy-website: release-content/0.15/release-notes/15800_Add_mesh_picking_backend_and_MeshRayCast_system_parameter.md

<!-- Add mesh picking backend and `MeshRayCast` system parameter -->
<!-- https://github.com/bevyengine/bevy/pull/15800 -->

> ![A collection of geometric shapes, with a pointer showing a point on a hovered mesh. The indicator is perpendicular to the surface.](mesh_picking.png)

Reviewer: It's a shame that mesh picking is called out further down as being disabled by default and discouraged due to performance concerns, but then the title image for this feature is explicitly mesh picking. I don't think there's anything that can be done about it though. These are real concerns, and mesh picking is the best visual representation of the feature.

Being able to click on objects to select them is a vital and seemingly simple task in any game.

> Since 2020, doing this in Bevy has largely meant pulling in `@aevyrie`'s beloved ecosystem crate, [`bevy_mod_picking`] and it's simple raycasting companion [`bevy_mod_raycast`].

Reviewer 2: ```suggestion
Since 2020, doing this in Bevy has largely meant pulling in `@aevyrie`'s beloved ecosystem crate, [`bevy_mod_picking`] and its simple raycasting companion [`bevy_mod_raycast`].
```

> Over the years, this crate has been refined and battle-tested, both by [Foresight Spatial Labs] (a CAD-creating, Bevy-using company co-founded by Aevyrie) and

Reviewer 2: I'm technically not a co-founder!
```suggestion
Over the years, this crate has been refined and battle-tested, both by [Foresight Spatial Labs] (where Aevyrie works) and
```

Bevy is thrilled to have had the chance to work with the team behind [`bevy_mod_picking`] and have adopted the project wholesale into Bevy itself.
Integrating a large project is a ton of work, and we're incredibly grateful to the contributors who have made `bevy_picking` a stable, first-class feature of the engine.

The new `bevy_picking` crate follows the existing modular architecture closely:

1. Inputs are gathered from mouse, touch and pen devices. Each pointing device (humans are equipped with 10 by default) gets a screen-space [`PointerLocation`].
2. Each modular [backend] performs the domain-specific work (like raycasting) of figuring out how these pointer locations map to [`PointerHits`] on objects that they're watching.
3. The hit information from each backend is combined and sorted to produce a coherent [`HoverMap`], which lists which entities each pointer is hovering over.
4. High level events (both ordinary events and observers!) are emitted for each hovered entity, capturing complex behavior such as clicking, dragging or releasing various objects.

> In Bevy 0.15, we're shipping with UI, sprite and mesh backends. Each of these comes with its own caveats for now:

Reviewer 2: ```suggestion
In Bevy 0.15, we're shipping three first-party picking backends for UI, sprites, and meshes. Each of these comes with its own caveats for now:
```

We expect both [`bevy_rapier`] and [`avian`] (the two most popular ecosystem physics crates for Bevy) to add their own accelerated collider picking backends to work with the newly upstreamed API. Unless you're debugging, building an editor or really care about the exact triangles of raw meshes, you should use one of those crates for efficient mesh picking.

## Usage

If you haven't used `bevy_picking`'s predecessor, there are two important and straightforward ways to get started with the API.

First, you might want to quickly update the state of your objects (be they UI or game objects) based on what is being done to them, typically highlighting them or changing their color. For that, simply match against the [`PickingInteraction`] component.

Secondly, you might want to respond dynamically to various pointer-powered events. For that, we recommend using observers (which replaced the existing `bevy_event_listener` solution during the upstreaming process).
Here, we're spawning a simple text node and responding to pointer events.

```rust
fn main() {
    use bevy::prelude::*;

>     App::new()
>         .add_plugins(DefaultPlugins)
>         .add_systems(Startup, setup)
>         .add_systems(Update, update_counter_display)
>         // Observers added globally will watch for events sent to *any* entity
>         .add_observer(change_text_color_on_hover)
>         .add_observer(reset_text_color)
>         .run();
> }
> 
> fn change_text_color_on_hover(
>     over: Trigger<Pointer<Over>>,
>     mut text_colors: Query<&mut TextColor>,
> ) {
>     if let Ok(mut text_color) = text_colors.get_mut(over.entity()) {
>         *text_color = Srgba::RED.into();
>     }
> }
> fn reset_text_color(out: Trigger<Pointer<Out>>, mut text_colors: Query<&mut TextColor>) {
>     if let Ok(mut text_color) = text_colors.get_mut(out.entity()) {
>         *text_color = TextColor::default();
>     }
> }
> #[derive(Component)]
> struct Counter(i32);
> fn setup(mut commands: Commands) {
>     commands.spawn(Camera2d::default());
>     // Root node
>     commands
>         .spawn(Node {
>             flex_direction: FlexDirection::Column,
>             ..Default::default()
>         })
>         .with_children(|builder| {
>             builder.spawn((Text::new("Counter: 0"), Counter(0)));
>             builder
>                 .spawn((Text::new("Count up :)"),))
>                 // Observers added to a single entity only watch for events to that specific entity
>                 .observe(
>                     |_click: Trigger<Pointer<Click>>,
>                      mut counter_query: Query<&mut Counter>| {
>                         if let Ok(mut counter) = counter_query.get_single_mut() {
>                             counter.0 += 1;
>                         }
>                     },
>                 );
>             builder.spawn((Text::new("Count down :("),)).observe(
>                 |_click: Trigger<Pointer<Click>>, mut counter_query: Query<&mut Counter>| {
>                     if let Ok(mut counter) = counter_query.get_single_mut() {
>                         counter.0 -= 1;
>                     }
>                 },
>             );
>         });
> }
> // Systems with the new `Single` system param are skipped if their query doesn't return exactly one elemnent
> fn update_counter_display(singleton_query: Single<(&mut Text, &Counter)>) {
>     let (mut text, counter) = singleton_query.into_inner();
>     *text = Text::new(format!("Counter: {}", counter.0));
> }

Reviewer 2: ```suggestion
use bevy::prelude::*;

fn main() {
    App::new()
        .add_plugins((DefaultPlugins, MeshPickingPlugin))
        .add_systems(Startup, setup_scene)
        .run();
}

fn setup_scene(
    mut commands: Commands,
    mut meshes: ResMut<Assets<Mesh>>,
    mut materials: ResMut<Assets<StandardMaterial>>,
) {
    // UI text that prints a message when clicked:
    commands
        .spawn((Text::new("Click Me!"), Node::default()))
        .observe(on_click_print_hello);

    // A cube that spins when dragged:
    commands
        .spawn((
            Mesh3d(meshes.add(Cuboid::default())),
            MeshMaterial3d(materials.add(Color::WHITE)),
        ))
        // Picking observers work with *any* entity that has a picking backend running.
        // Try adding this `on_drag_spin` observer to the UI text! :)
        .observe(on_drag_spin);

    // Light and camera
    commands.spawn((PointLight::default(), Transform::from_xyz(4.0, 8.0, 4.0)));
    commands.spawn((Camera3d::default(), Transform::from_xyz(0.0, 2.0, 9.0)));
}

fn on_click_print_hello(click: Trigger<Pointer<Click>>) {
    println!("{} was clicked!", click.entity());
}

fn on_drag_spin(drag: Trigger<Pointer<Drag>>, mut transforms: Query<&mut Transform>) {
    let mut transform = transforms.get_mut(drag.entity()).unwrap();
    transform.rotate_y(drag.delta.x * 0.02);
}
```

I'd like to propose a simpler example that demonstrates a few things I want to highlight:

- picking works for anything that has a backend running
- picking observers look a lot like systems
- the example encourages you to play with it, to show you how observers are not "special" for the type of entity, if it supports picking, you can respond to picking triggers. Being able to add a "spin on drag" observer to text *and* a mesh, and have it work, is pretty incredible. I think it's a great example of how ECS allows for powerful composition.

I know this might seem like the nittiest of nits, but I think this is something that makes the feature really stand out, and I want people to see that it's not just about UI, it's about interacting with all entities.

Also, without enter/leave, the current example will unfortunately guide users into the over/out footgun. I do appreciate what it is doing though!

> ```

Reviewer: I think this example is a bit lengthy without any explanation in-between. Even though it glosses over some details, I think breaking it up to just focus on hovering and clicking separately can help:

_I wanted to post the below as a suggestion but GitHub doesn't like the multiple code blocks inside a suggestion block._

For example, you can use `Pointer<Over>` and `Pointer<Out>` to highlight entities under the pointer.

```rust
fn change_text_color_on_hover(
    over: Trigger<Pointer<Over>>,
    mut text_colors: Query<&mut TextColor>,
) {
    if let Ok(mut text_color) = text_colors.get_mut(over.entity()) {
        *text_color = Srgba::RED.into();
    }
}

fn reset_text_color(
    out: Trigger<Pointer<Out>>,
    mut text_colors: Query<&mut TextColor>,
) {
    if let Ok(mut text_color) = text_colors.get_mut(out.entity()) {
        *text_color = TextColor::default();
    }
}
```

You could also `Pointer<Click>` to let users interact with entities.

```rust
#[derive(Component)]
struct Counter(i32);

fn setup(mut commands: Commands) {
    commands
        .spawn(Node {
            flex_direction: FlexDirection::Column,
            ..Default::default()
        })
        .with_children(|builder| {
            builder.spawn((Text::new("Counter: 0"), Counter(0)));
            builder
                .spawn((Text::new("Count up :)"),))
                // Observers added to a single entity only watch for events to that specific entity
                .observe(
                    |_click: Trigger<Pointer<Click>>,
                     mut counter_query: Query<&mut Counter>| {
                        if let Ok(mut counter) = counter_query.get_single_mut() {
                            counter.0 += 1;
                        }
                    },
                );
        });
}
```

> If you want to control how an entity interacts with picking, add the [`PickingBehavior`] component to them and configure it to meet your needs.

Reviewer 2: ```suggestion
If you want to override how an entity interacts with picking, add the [`PickingBehavior`] component to them and configure it to meet your needs.
```

1. Inputs are gathered from mouse, touch and pen devices. Each pointing device (humans are equipped with 10 by default) gets a screen-space [`PointerLocation`].
2. Each modular [backend] performs the domain-specific work (like raycasting) of figuring out how these pointer locations map to [`PointerHits`] on objects that they're watching.
3. The hit information from each backend is combined and sorted to produce a coherent [`HoverMap`], which lists which entities each pointer is hovering over.
4. High level events (both ordinary events and observers!) are emitted for each hovered entity, capturing complex behavior such as clicking, dragging or releasing various objects.

In Bevy 0.15, we're shipping with UI, sprite and mesh backends. Each of these comes with its own caveats for now:

- UI: both the legacy [`Interaction`] and new [`PickingInteraction`] components exist [for now](https://github.com/bevyengine/bevy/issues/15550), with subtle behavioral differences.
- Sprites: picking always uses the full rectangle, and [alpha transparency is not taken into account](https://github.com/bevyengine/bevy/issues/14929).

> - Mesh: this is a naive raycast against the full mesh, and can be painfully slow. You should be using simplified meshes and an acceleration data structure like a BVH to speed this up if you are using this functionality for games. As a result, this functionality is disabled by default.

Reviewer 3: I probably wouldn't say "painfully slow", just "relatively slow" or even "slow for large scenes with lots of geometry". There's still AABB culling and a lot of parallelism, and even without a BVH, it was taking just 0.34% of frame time for the Caldera hotel in my quick (not very thorough) [test](https://github.com/bevyengine/bevy/pull/15800#issuecomment-2408999324). Of course that's just a single ray per frame though.

But yeah, it's also important to make it clear that it's not intended as a replacement for proper ray casting based on colliders, and it should mainly be used for debugging, editor use cases, or simple applications that don't have too much geometry.

Author: ```suggestion
- Mesh: this is a naive raycast against the full mesh, and can be relatively slow. You should be using simplified meshes and an acceleration data structure like a BVH to speed this up if you are using this functionality for performance-sensitive games. As a result, this functionality is disabled by default.
```

> We expect both [`bevy_rapier`] and [`avian`] (the two most popular ecosystem physics crates for Bevy) to update their accelerated mesh picking implementations to work with the newly upstreamed API. Unless you're debugging, building an editor or really care about the exact triangles of raw meshes, you should use one of those crates for efficient mesh picking.

Reviewer 3: They don't have "mesh picking" (prior to 0.15), just spatial queries / scene queries, which includes ray casting for colliders (not strictly meshes).
```suggestion
We expect both [`bevy_rapier`] and [`avian`] (the two most popular ecosystem physics crates for Bevy) to add their own accelerated collider picking backends to work with the newly upstreamed API. Unless you're debugging, building an editor or really care about the exact triangles of raw meshes, you should use one of those crates for efficient mesh picking.
```
Also, I'm not sure if it's worth mentioning here since it's not released yet, but Avian already has the picking backend merged on the main branch (see https://github.com/Jondolf/avian/pull/554). No idea on bevy_rapier's plans/timeline here

1. Inputs are gathered from mouse, touch and pen devices. Each pointing device (humans are equipped with 10 by default) gets a screen-space [`PointerLocation`].
2. Each modular [backend] performs the domain-specific work (like raycasting) of figuring out how these pointer locations map to [`PointerHits`] on objects that they're watching.
3. The hit information from each backend is combined and sorted to produce a coherent [`HoverMap`], which lists which entities each pointer is hovering over.
4. High level events (both ordinary events and observers!) are emitted for each hovered entity, capturing complex behavior such as clicking, dragging or releasing various objects.

In Bevy 0.15, we're shipping three first-party picking backends for UI, sprites, and meshes. Each of these comes with its own caveats for now:

- UI: both the legacy [`Interaction`] and new [`PickingInteraction`] components exist [for now](https://github.com/bevyengine/bevy/issues/15550), with subtle behavioral differences.
- Sprites: picking always uses the full rectangle, and [alpha transparency is not taken into account](https://github.com/bevyengine/bevy/issues/14929).

> - Mesh: this is a naive raycast against the full mesh, and can be relatively slow. You should be using simplified meshes and an acceleration data structure like a BVH to speed this up if you are using this functionality for performance-sensitive games. As a result, this functionality is disabled by default.

Author: ```suggestion
- Mesh: this is a naive raycast against the full mesh. If you run into performance problems here, you should use simplified meshes and an acceleration data structure like a BVH to speed this up. As a result, this functionality is currently disabled by default.
```

## Usage

If you haven't used `bevy_picking`'s predecessor, there are two important and straightforward ways to get started with the API.

First, you might want to quickly update the state of your objects (be they UI or game objects) based on what is being done to them, typically highlighting them or changing their color. For that, simply match against the [`PickingInteraction`] component.

Secondly, you might want to respond dynamically to various pointer-powered events. For that, we recommend using observers (which replaced the existing `bevy_event_listener` solution during the upstreaming process).
Here, we're spawning a simple text node and responding to pointer events.

```rust
use bevy::prelude::*;

fn main() {
    App::new()
        .add_plugins((DefaultPlugins, MeshPickingPlugin))
        .add_systems(Startup, setup_scene)
        .run();
}

fn setup_scene(
    mut commands: Commands,
    mut meshes: ResMut<Assets<Mesh>>,
    mut materials: ResMut<Assets<StandardMaterial>>,
) {
    // UI text that prints a message when clicked:
    commands

>         .spawn((Text::new("Click Me!"), Node::default()))

Reviewer 4: `Text` requires `Node`, so it doesn't need to be specified manually 

```suggestion
        .spawn(Text::new("Click Me!"))
```

Reviewer 5: I suspect this is something people will run into a lot. It’s not a big issue, but perhaps we can lint/warn against initializing a default component that is also a requirement of another component in the list?
